From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: marcost2 <marcostolcachir@gmail.com>
Date: Sat, 21 Sep 2024 15:30:35 -0300
Subject: [PATCH] techpack: display: Expose fod_hbm as sysfs

* Based on cosmin's work for EQS, adapted to moto's interfaces on here

Original patch:
https://github.com/marcost2/android_kernel_motorola_sm6375/commit/14cf1d3785f913743fc568e6c9d732de6be06cf1

Signed-off-by: Sevenrock <sevenrock@hotmail.de>
---
 techpack/display/msm/dsi/dsi_panel.c | 81 ++++++++++++++++++++++++++++
 techpack/display/msm/dsi/dsi_panel.h |  1 +
 2 files changed, 82 insertions(+)

diff --git a/techpack/display/msm/dsi/dsi_panel.c b/techpack/display/msm/dsi/dsi_panel.c
index 27e95fa6999d..a3a3b59f8cc9 100644
--- a/techpack/display/msm/dsi/dsi_panel.c
+++ b/techpack/display/msm/dsi/dsi_panel.c
@@ -1347,6 +1347,81 @@ static int dsi_panel_set_hbm(struct dsi_panel *panel,
 	return rc;
 };
 
+static ssize_t sysfs_fod_hbm_read(struct device *dev,
+				  struct device_attribute *attr,
+				  char *buf)
+{
+	struct dsi_display *display = dev_get_drvdata(dev);
+	struct dsi_panel *panel = display->panel;
+	bool status;
+
+	mutex_lock(&panel->panel_lock);
+	status = panel->fod_hbm_enabled;
+	mutex_unlock(&panel->panel_lock);
+
+	return snprintf(buf, PAGE_SIZE, "%u\n", status);
+}
+
+static ssize_t sysfs_fod_hbm_write(struct device *dev,
+				   struct device_attribute *attr,
+				   const char *buf, size_t count)
+{
+	struct dsi_display *display = dev_get_drvdata(dev);
+	struct dsi_panel *panel = display->panel;
+	struct msm_param_info param_info;
+	bool status;
+	int rc;
+
+	if (!panel->lhbm_config.enable)
+		return -EOPNOTSUPP;
+
+	rc = kstrtobool(buf, &status);
+	if (rc)
+		return rc;
+
+	if (panel->fod_hbm_enabled == status)
+		goto exit;
+
+	param_info.param_idx = PARAM_HBM_ID;
+	param_info.value = status ? HBM_FOD_ON_STATE : HBM_OFF_STATE;
+	rc = dsi_panel_set_hbm(panel, &param_info);
+	if (rc)
+		goto exit;
+
+	panel->fod_hbm_enabled = status;
+
+exit:
+
+	return rc ?: count;
+}
+
+static DEVICE_ATTR(fod_hbm, 0644, sysfs_fod_hbm_read, sysfs_fod_hbm_write);
+
+static struct attribute *panel_attrs[] = {
+	&dev_attr_fod_hbm.attr,
+	NULL,
+};
+
+static struct attribute_group panel_attrs_group = {
+	.attrs = panel_attrs,
+};
+
+static int dsi_panel_sysfs_init(struct dsi_panel *panel)
+{
+	int rc;
+
+	rc = sysfs_create_group(&panel->parent->kobj, &panel_attrs_group);
+	if (rc)
+		DSI_ERR("failed to create panel sysfs attributes\n");
+
+	return rc;
+}
+
+static void dsi_panel_sysfs_deinit(struct dsi_panel *panel)
+{
+	sysfs_remove_group(&panel->parent->kobj, &panel_attrs_group);
+}
+
 static int dsi_panel_set_acl(struct dsi_panel *panel,
                         struct msm_param_info *param_info)
 {
@@ -4907,6 +4982,10 @@ struct dsi_panel *dsi_panel_get(struct device *parent,
 
 	mutex_init(&panel->panel_lock);
 
+	rc = dsi_panel_sysfs_init(panel);
+	if (rc)
+		goto error;
+
 	return panel;
 error_vreg_put:
 	(void)dsi_panel_vreg_put(panel);
@@ -4917,6 +4996,8 @@ struct dsi_panel *dsi_panel_get(struct device *parent,
 
 void dsi_panel_put(struct dsi_panel *panel)
 {
+	dsi_panel_sysfs_deinit(panel);
+
 	drm_panel_remove(&panel->drm_panel);
 
 	/* free resources allocated for ESD check */
diff --git a/techpack/display/msm/dsi/dsi_panel.h b/techpack/display/msm/dsi/dsi_panel.h
index 390bc3049660..c989468b1489 100644
--- a/techpack/display/msm/dsi/dsi_panel.h
+++ b/techpack/display/msm/dsi/dsi_panel.h
@@ -420,6 +420,7 @@ struct dsi_panel {
 
 	bool is_hbm_using_51_cmd;
 	int hbm_en_gpio;
+	bool fod_hbm_enabled;
 };
 
 bool dsi_display_all_displays_dead(void);
-- 
2.39.5

